File: 2024/leet_code/daily_temperatures.java
public class Tuple {

  public final int idx;
  public final int val;

  public Tuple(int idx, int val) {
    this.idx = idx;
    this.val = val;
  }
}

class Solution {

  public int[] dailyTemperatures(int[] temperatures) {
    Stack<Tuple> stack = new Stack<Tuple>();
    int[] res = new int[temperatures.length];
    int curr;
    Tuple top;

    for (int i = 1; i < temperatures.length; i++) {
      top = new Tuple(i - 1, temperatures[i - 1]);
      stack.add(top);
      curr = temperatures[i];
      while (!stack.empty() && curr > stack.peek().val) {
        top = stack.pop();
        res[top.idx] = i - top.idx;
      }
    }
    return res;
  }
}

File: 2024/leet_code/daily_temperatures.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # identical to next-greatest-element implementation, except store index differential in result array instead of n.g.e.
        stack = []
        res = [0 for _ in range(len(temperatures))]
        for i in range(1, len(temperatures)):
            stack.append((temperatures[i - 1], i - 1))  # push to the stack
            curr = temperatures[i]
            while len(stack) > 0 and curr > stack[-1][0]:  # peek the top of the stack
                _, idx = stack.pop()
                res[idx] = i - idx
        return res
